File: lib/features/nilai/domain/nilai.dart
class Nilai {
  final String uid, nim, nama;
  final int tugas, uts, uas;

  const Nilai({
    required this.uid,
    required this.nim,
    required this.nama,
    this.tugas = 0,
    this.uts = 0,
    this.uas = 0,
  });

  factory Nilai.fromMap(Map<String, dynamic> map) => Nilai(
        uid: map['mahasiswa_uid'] ?? map['uid'] ?? '',
        nim: map['nim'] ?? '',
        nama: map['nama'] ?? '',
        tugas: (map['tugas'] as num?)?.toInt() ?? 0,
        uts: (map['uts'] as num?)?.toInt() ?? 0,
        uas: (map['uas'] as num?)?.toInt() ?? 0,
      );

  double get nilaiAkhir => (tugas * 0.3) + (uts * 0.3) + (uas * 0.4);

  String get grade {
    final n = nilaiAkhir;
    if (n >= 85) return 'A';
    if (n >= 75) return 'B';
    if (n >= 65) return 'C';
    if (n >= 55) return 'D';
    return 'E';
  }
}

File: app/Models/Nilai.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Nilai extends Model
{
    protected $table = 'nilai';
    public $timestamps = false;

    protected $fillable = [
        'kelas',
        'kelas_kuliah_id',
        'mahasiswa_uid',
        'nim',
        'nama',
        'tugas',
        'uts',
        'uas',
    ];

    public function mahasiswa(): BelongsTo
    {
        return $this->belongsTo(Mahasiswa::class, 'mahasiswa_uid', 'uid');
    }

    public function kelasKuliah(): BelongsTo
    {
        return $this->belongsTo(KelasKuliah::class, 'kelas_kuliah_id', 'id');
    }

    public function getNilaiAkhirAttribute(): float
    {
        return ($this->tugas * 0.3) + ($this->uts * 0.3) + ($this->uas * 0.4);
    }

    public function getGradeAttribute(): string
    {
        $n = $this->nilai_akhir;
        if ($n >= 85) return 'A';
        if ($n >= 75) return 'B';
        if ($n >= 65) return 'C';
        if ($n >= 55) return 'D';
        return 'E';
    }
}
